### ISSUE TITLE
Separate enums and constants from the rest of the generated client

### ISSUE THREAD

**Issue by cawa-93** [prisma/prisma#13567]
## Problem

You cannot easily import enum in client code.
```ts
// src/app.ts
import { MyEnum } from '@prisma/client'
console.log(MyEnum)
```
The problem is that all enums are in the same file with the rest of the prism code. This leads to two problems for bundlers:
- Tree-shacking. To import a single enum, you need to analyse and import the entire Prisma dependency tree. Which is not without its side effects. Because of this, shaking the tree is impossible.
- https://github.com/prisma/prisma/issues/12504#issuecomment-1136126199

## Suggested solution

Divide the generated client file into several separate ones.
```js
// .prisma/client/index.js
// ...
exports.Prisma = Prisma
exports.MyEnum = required('./enums').MyEnum
```
```js
//  .prisma/client/enums.js
function makeEnum(x) { return x; }

exports.MyEnum = makeEnum({ /* ... */ })
// will be tree-shaked
exports.AnotherEnum = makeEnum({ /* ... */ })
```
```js
// @prisma/client/enums
module.exports = {
  ...require('.prisma/client/enums'),
}
```

This allows you import single enum from separate file
```ts
import { MyEnum } from '@prisma/client/enums'
```

That will not require the bundler to analyze the entire dependency tree. This will allow you to import enums into client code and work around this issue

**Comment by shtse8** [comment]
Support. Hope this can be addressed soon.

**Comment by therealokoro** [comment]
Any update on this @jkomyno  and @janpio

**Comment by janpio** [comment]
No, we will post an update here when we get to work on this or have implemented it. Until then, it is an improvement request here in our issues and others how have the same problem, or want to explain their use case, are welcome to do so.

**Comment by bennettdams** [comment]
I have the same problem, in my case with a Next.js app. 

Additionally - and I haven't tried this out yet - is this maybe also true for Prisma's `validator`?
You cannot destructure the import, so you'll always have to import `Prisma`:


```ts
import { Prisma} from '@prisma/client'

const foo = Prisma.validator(...)
```

**Comment by bennettdams** [comment]
## Workaround

For the each enum model, Prisma does not only give you a `const` from `@prisma/client` but also the type as string literal, for example:

![image](https://user-images.githubusercontent.com/29319414/204787684-e4a6f381-1574-49e6-b534-6d071ab648fb.png)

..which means you can force only importing the type easily:

`import type { PostSegmentImagePosition } from '@prisma/client'`

You can now use the string literal directly, or, if you want to use them in an "enum way", you can build your own:

```ts
const imagePositionEnum: { [K in PostSegmentImagePosition]: null } = {
  RIGHT: null,
  BOTTOM: null,
}
```

This will give you the typical autocomplete:


![image](https://user-images.githubusercontent.com/29319414/204788306-5d04d49c-f8f0-4ad4-a063-543f2646022d.png)

The good thing about this that TypeScript will yell at you if you forget an enum in your definition:

![image](https://user-images.githubusercontent.com/29319414/204788829-f0881f28-4e9f-43e5-b51c-2fd223c7059e.png)

Using this workaround fixes `@prisma/client` being part of the bundle as you only import the type and not the constant.

**Comment by victorshx** [comment]
@jkomyno and @janpio Any update on this? It's affecting Vite builds, it should be on the top priority list.

**Comment by juliusmarminge** [comment]
We managed to reduce cal.com bundle size by 5% using an enum generator to import enums from a separate entrypoint. I made a gist here:

https://gist.github.com/juliusmarminge/b06a3e421117a56ba1fea54e2a4c0fcb

**Comment by tikoflano** [comment]
Adding another use case for the separate enums file.

I'm using NextJS 14 and Prisma 5.8. I'm using the  [prisma-field-encryption](https://www.npmjs.com/package/prisma-field-encryption) client extension which uses Node's **crypto** module.

The problem with this is that when importing my prisma client, it runs `require("prisma-field-encryption")` to load the extension. As this extension uses the Node's **crypto** it fails when I add `import { MyEnum } from '@prisma/client'` from a **client component**, because can't access Node's modules from the client runtime. Why I import the Prisma client form a client component? Because I want to access MyEnum to compare values.

The workaround is to export the enums from a separate file but:

1. The export enums file needs to be updated every time we add a new enum to the schema (easy to forget)
2. More importan, when using the MyEnum, Intellisense auto imports it from '@prisma/client' and making it fail. The team can easily forget that they need to import them manually from the separate enums file.

**Comment by Dutchgoose** [comment]
Would love to see this implemented!

**Comment by gruckion** [comment]
Please address this issue prisma.

**Comment by TannerPerrien** [comment]
For a clean workaround to get you through for the time being, check out [prisma-enum-generator](https://github.com/ajmnz/prisma-enum-generator). The only downside I have encountered so far is that enums get generated in two places and auto-imports in your IDE may import from your prisma package (easy fix).

See [juliusmarminge](https://github.com/juliusmarminge)'s earlier comment on this issue pointing to a solution of the same nature.

**Comment by rhyek** [comment]
> For a clean workaround to get you through for the time being, check out [prisma-enum-generator](https://github.com/ajmnz/prisma-enum-generator). The only downside I have encountered so far is that enums get generated in two places and auto-imports in your IDE may import from your prisma package (easy fix).
> 
> See [juliusmarminge](https://github.com/juliusmarminge)'s earlier comment on this issue pointing to a solution of the same nature.

Sadly, this doesn't address the issue. This is generating different enums than those you specify in your prisma schema. It's for a different purpose.

**Comment by michaelhays** [comment]
I believe this is fixed with the `prisma-client` generator as of Prisma 6.7.0: https://www.prisma.io/docs/orm/prisma-schema/overview/generators#output-splitting-and-importing-types
